homeassistant/components/homekit_controller.py

```diff
@@ -0,0 +1,178 @@
+
+    class hk_device(Entity):
+        """Homekit device"""
+
+        def __init__(self, hass, host, port, model, hkid, config_num, config):
+            import homekit
+
+            _LOGGER.info("Setting up Homekit device %s", model)
+            self.host = host
+            self.port = port
+            self.model = model
+            self.hkid = hkid
+            self.config_num = config_num
+            self.config = config
+            self.configurator = hass.components.configurator
+
+            data_dir = os.path.join(get_default_config_dir(), homekit_dir)
+            if not os.path.isdir(data_dir):
+                os.mkdir(data_dir)
+            self.pairing_file = os.path.join(data_dir, "hk-" + hkid)
+            self.pairing_data = homekit.load_pairing(self.pairing_file)
+
+            # Monkey patch httpclient for increased compatibility
+            http.client.HTTPConnection._send_output = homekit_http_send
+
+            self.conn = http.client.HTTPConnection(self.host, port=self.port)
+            if self.pairing_data is not None:
+                self.accessory_setup()
+            else:
+                self.configure()
+
+        def get_serial(self, accessory):
+            for service in accessory['services']:
+                if homekit.ServicesTypes.get_short(service['type']) != 'accessory-information':
+                    continue
+                for characteristic in service['characteristics']:
+                    ctype = homekit.CharacteristicsTypes.get_short(
+                        characteristic['type'])
+                    if ctype != 'serial-number':
+                        continue
+                    return characteristic['value']
+            return None
+
+        def accessory_setup(self):
+            self.controllerkey, self.accessorykey = \
+                homekit.get_session_keys(self.conn, self.pairing_data)
+            self.securecon = homekit.SecureHttp(self.conn.sock,
+                                                self.accessorykey,
+                                                self.controllerkey)
+            response = self.securecon.get('/accessories')
+            data = json.loads(response.read().decode())
+            for accessory in data['accessories']:
+                serial = self.get_serial(accessory)
+                if serial in KNOWN_ACCESSORIES:
+                    continue
+                aid = accessory['aid']
+                for service in accessory['services']:
+                    s = {'type': homekit.ServicesTypes.get_short(service['type']),
+                         'aid': aid,
+                         'service': service,
+                         'device': self}
+                    _LOGGER.info("Found %s", s['type'])
+                    component = HOMEKIT_ACCESSORY_DISPATCH.get(s['type'], None)
+                    if component is not None:
+                        discovery.load_platform(hass, component, DOMAIN, s,
```

Comment:
Discovery info is only allowed to contain JSON serializable values as it leverages the event bus. So instead of sending all info over, send over the aid/service combo and fetch it from `hass.data` inside the platform setup.

Reply:
Done